Tools/scripts/deepfreeze.py

```diff
@@ -0,0 +1,421 @@
+        maxint = sys.maxsize
+        if maxint == 2**31 - 1:
+            digit = 2**15
+        elif maxint == 2**63 - 1:
+            digit = 2**30
+        else:
+            assert False, f"What int size is this system?!? {maxint=}"
+        sign = -1 if i < 0 else 0 if i == 0 else +1
+        i = abs(i)
+        digits: list[int] = []
+        while i:
+            i, rem = divmod(i, digit)
+            digits.append(rem)
+        self.write("static")
+        with self.indent():
+            with self.block("struct"):
+                self.write("PyObject_VAR_HEAD")
+                self.write(f"digit ob_digit[{max(1, len(digits))}];")
+        with self.block(f"{name} =", ";"):
+            self.object_var_head("PyLong_Type", sign*len(digits))
+            if digits:
+                ds = ", ".join(map(str, digits))
+                self.write(f".ob_digit = {{ {ds} }},")
+        return f"& {name}.ob_base.ob_base"
+
+    def generate_float(self, name: str, x: float) -> str:
+        with self.block(f"static PyFloatObject {name} =", ";"):
+            self.object_head("PyFloat_Type")
+            self.write(f".ob_fval = {x},")
+        return f"&{name}.ob_base"
+
+    def generate_complex(self, name: str, z: complex) -> str:
+        with self.block(f"static PyComplexObject {name} =", ";"):
+            self.object_head("PyComplex_Type")
+            self.write(f".cval = {{ {z.real}, {z.imag} }},")
+        return f"&{name}.ob_base"
+
+    def generate_frozenset(self, name: str, fs: frozenset[object]) -> str:
+        ret = self.generate_tuple(name, tuple(sorted(fs)))
+        self.write("// TODO: The above tuple should be a frozenset")
+        return ret
+
+    def generate(self, name: str, obj: object) -> str:
+        # Use repr() in the key to distinguish -0.0 from +0.0
+        key = (type(obj), obj, repr(obj))
+        if key in self.cache:
+            self.hits += 1
+            # print(f"Cache hit {key!r:.40}: {self.cache[key]!r:.40}")
+            return self.cache[key]
+        self.misses += 1
+        match obj:
+            case types.CodeType() as code:
+                val = self.generate_code(name, code)
+            case tuple(t):
+                val = self.generate_tuple(name, t)
+            case str(s):
+                val = self.generate_unicode(name, s)
+            case bytes(b):
+                val = self.generate_bytes(name, b)
+            case True:
+                return "Py_True"
+            case False:
+                return "Py_False"
+            case int(i):
+                val = self.generate_int(name, i)
+            case float(x):
+                val = self.generate_float(name, x)
+            case complex() as z:
+                val = self.generate_complex(name, z)
+            case frozenset(fs):
+                val = self.generate_frozenset(name, fs)
+            case builtins.Ellipsis:
+                return "Py_Ellipsis"
+            case None:
+                return "Py_None"
```

Comment:
Interesting.  I hadn't realized that marshal doesn't explicitly handle the one remaining singleton,`NotImplemented`.

Reply:
That's presumably because it never occurs as a constant in a code object.